### PR TITLE
Handle TTD and mining pre and post merge 

### LINKED ISSUE #2898: Merge: Consistently Handle Terminal Total Difficulty
### Description
When the terminal proof-of-work block is reached ensure we:
 * stop any mining activities
 * disable sending and handling of NEW_BLOCK_HASHES and NEW_BLOCK messages 
 * stop sync process

TTD is part of genesis, but should be override-able via the CLI 

### Acceptance Criteria
* correct behavior at TTD for the above cases
* correct behavior for a re-org below TTD
* test coverage
* create a terminal-total-difficulty CLI override parameter which overrides the TTD from genesis config file

### Additional Information
Much of this behavior is implemented already, but there are corner cases and no test coverage

### PR DESCRIPTION
**Pull request by garyschulte**
Signed-off-by: garyschulte <garyschulte@gmail.com>

## PR description
Correctly and consistently handle mining pre/post merge for PoW and Clique networks.  The majority of this PR is code from the merge branch which needed some 'treatment' in order to not cause regressions in behavior on main.

* for clarity, rename isStoppedAtTerminalDifficulty to hasReachedTerminalDifficulty 
* use AtomicReference for PostMergeContext rather than synchronized
* add mining parameter override for consensus mechanisms that mine regardless of miningParameters
* merge parameter validation from `merge` branch which allows mining parameters without PoW mining (if merge is enabled)
* merge fix for optional in TxPoolEvictionService
* fix for ancestorIsValidTerminalProofOfWork in MergeCoordinator when close to TTD
* address Rayonism tech-debt, use setBlockChoiceRule at TTD rather than disabling reorg behavior entirely pre-merge

## Fixed Issue(s)

fixes #2898 

## Changelog

- [ ] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/PostMergeContext.java [hunk 1]
```diff
@@ -73,35 +71,38 @@ public static synchronized PostMergeContext get() {
...
+      throw new IllegalStateException("cannot set null terminal total difficulty");
+    }
     terminalTotalDifficulty.set(newTerminalTotalDifficulty);
     return this;
   }
 
   @Override
   public void setIsPostMerge(final Difficulty totalDifficulty) {
-    if (Optional.ofNullable(lastFinalized.get()).isPresent()) {
+    if (isPostMerge.get().orElse(Boolean.FALSE)
```
Comment (fab-10) [inline consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/PostMergeContext.java:83]: my understanding is that if we have a finalized block, then we are post merge, so this check seems redundant, am I missing something?

**Review by fab-10 — APPROVED** [review]
LGTM, just a comment on a possible redundant check

File: consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/PostMergeContext.java (hunk 1 above)
Comment (garyschulte) [reply consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/PostMergeContext.java:83]: Yes the finalized check is unnecessary now.  The post-merge condition is 'stronger' - we will never switch back to the prior consensus mechanism and will instead rely on forkchoiceUpdated to direct any reorg around TTD.  

I will remove the unnecessary finalized check and update the comment 👍
